modules/javafx.base/src/main/java/com/sun/javafx/binding/ListenerManager.java

```diff
@@ -0,0 +1,189 @@
+     * @param listener a listener to add, cannot be {@code null}
+     * @throws NullPointerException when listener is {@code null}
+     */
+    public void addListener(I instance, InvalidationListener listener) {
+        Objects.requireNonNull(listener);
+
+        instance.getValue();  // always trigger validation when adding an invalidation listener (required by tests)
+
+        Object data = getData(instance);
+
+        if (data == null) {
+            setData(instance, listener);
+        }
+        else if (data instanceof ListenerList<?> list) {
+            list.add(listener);
+        }
+        else {
+            setData(instance, new ListenerList<>(data, listener));
+        }
+    }
+
+    /**
+     * Adds a change listener.
+     *
+     * @param instance the instance to which the listeners belong, cannot be {@code null}
+     * @param listener a listener to add, cannot be {@code null}
+     * @throws NullPointerException when listener is {@code null}
+     */
+    public void addListener(I instance, ChangeListener<? super T> listener) {
+        Objects.requireNonNull(listener);
+
+        instance.getValue();  // always trigger validation when adding a change listener (required by tests)
+
+        Object data = getData(instance);
+
+        if (data == null) {
+            setData(instance, listener);
+        }
+        else if (data instanceof ListenerList<?> list) {
+            list.add(listener);
+        }
+        else {
+            setData(instance, new ListenerList<>(data, listener));
+        }
+    }
+
+    /**
+     * Removes a listener.
+     *
+     * @param instance the instance to which the listeners belong, cannot be {@code null}
+     * @param listener a listener to remove, cannot be {@code null}
+     * @return {@code true} if there are no more listeners registered after this call completes, otherwise {@code false}
+     * @throws NullPointerException when listener is {@code null}
+     */
+    public boolean removeListener(I instance, Object listener) {
+        Objects.requireNonNull(listener);
+
+        Object data = getData(instance);
+
+        if (data == null) {
+            return true;
+        }
+
+        if (data.equals(listener)) {
+            setData(instance, null);
+
+            return true;
+        }
+
+        if (data instanceof ListenerList<?> list) {
+            list.remove(listener);
+
+            updateAfterRemoval(instance, list);
+
+            return list.totalListeners() == 0;
+        }
+
+        return false;
+    }
+
+    /**
+     * Notifies the listeners managed in the given instance.
+     *
+     * @param instance the instance to which the listeners belong, cannot be {@code null}
+     * @param oldValue the previous value before this change occurred, can be {@code null}
+     */
+    public void fireValueChanged(I instance, T oldValue) {
+        Object data = getData(instance);
```

Comment:
The `data` value could be passed into this method, which would save a (potentially not devirtualized) method call.

Reply:
Thanks, I'll look into that, it might speed up the 1 listener cases a bit.  The same applies to OldValueCachingListenerManager#getValue I think.  I know it isn't possible for the add/remove calls, as the data may change if they're nested, but for `fireValueChanged` I never really checked after going to this strategy.

Reply:
Have you considered passing `data` directly into the method? What is your conclusion?